libethereum/ChainParams.cpp

```diff
@@ -49,44 +49,67 @@ ChainParams::ChainParams()
 
 ChainParams::ChainParams(std::string const& _json, h256 const& _stateRoot)
 {
+	try
+	{
+		loadConfig(_json, _stateRoot);
+	}
+	catch (...)
+	{
+		cerr << "Bad json file provided. Should be a config file" << endl;
+	}
+}
+
+void ChainParams::loadConfig(std::string const& _json, h256 const& _stateRoot)
+{
 	js::mValue val;
 	json_spirit::read_string(_json, val);
 	js::mObject obj = val.get_obj();
+
 	sealEngineName = obj["sealEngine"].get_str();
-	js::mObject genesis = obj["genesis"].get_obj();
+	// params
 	js::mObject params = obj["params"].get_obj();
+	accountStartNonce = u256(fromBigEndian<u256>(fromHex(params["accountStartNonce"].get_str())));
+	maximumExtraDataSize = u256(fromBigEndian<u256>(fromHex(params["maximumExtraDataSize"].get_str())));
+	tieBreakingGas = params.count("tieBreakingGas") ? params["tieBreakingGas"].get_bool() : true;
+	blockReward = u256(fromBigEndian<u256>(fromHex(params["blockReward"].get_str())));
+	for (auto i: params)
+		if (i.first != "accountStartNonce" && i.first != "maximumExtraDataSize" && i.first != "blockReward" && i.first != "tieBreakingGas")
+			otherParams[i.first] = i.second.get_str();
+	// genesis
+	string jj = obj["genesis"].get_value().get_str();
+	loadGenesis(jj, _stateRoot);
+	// genesis state
+	loadGenesisState(obj["accounts"].get_str());
+}
 
-	// params
-	{
-		accountStartNonce = u256(fromBigEndian<u256>(fromHex(params["accountStartNonce"].get_str())));
-		maximumExtraDataSize = u256(fromBigEndian<u256>(fromHex(params["maximumExtraDataSize"].get_str())));
-		tieBreakingGas = params.count("tieBreakingGas") ? params["tieBreakingGas"].get_bool() : true;
-		blockReward = u256(fromBigEndian<u256>(fromHex(params["blockReward"].get_str())));
-		for (auto i: params)
-			if (i.first != "accountStartNonce" && i.first != "maximumExtraDataSize" && i.first != "blockReward" && i.first != "tieBreakingGas")
-				otherParams[i.first] = i.second.get_str();
-	}
+void ChainParams::loadGenesisState(std::string const& _json, std::unordered_map<Address, PrecompiledContract> const& _precompiled)
+{
+	precompiled = _precompiled;
+	genesisState = jsonToAccountMap(_json, nullptr, &precompiled);
```

Comment:
If this is not done for `loadGenesis`, does it initialise the account balances of the genesis block?